rs-matter/src/data_model/networks/wireless/thread.rs

```diff
@@ -0,0 +1,207 @@
+/*
+ *
+ *    Copyright (c) 2020-2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+//! This module contains Thread-specific types.
+
+use core::fmt::{Debug, Display};
+
+use crate::error::{Error, ErrorCode};
+use crate::tlv::{FromTLV, OctetsOwned, ToTLV};
+use crate::utils::init::{init, Init, IntoFallibleInit};
+use crate::utils::storage::Vec;
+
+use crate::data_model::sdm::net_comm::WirelessCreds;
+
+use super::{WirelessNetwork, WirelessNetworks};
+
+pub type ThreadNetworks<const N: usize, M> = WirelessNetworks<N, M, Thread>;
+
+/// A struct implementing the `WirelessNetwork` trait for Thread networks.
+#[derive(Debug, Clone, Eq, PartialEq, Hash, ToTLV, FromTLV)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct Thread {
+    /// Thread dataset in TLV format
+    pub dataset: OctetsOwned<256>,
+}
+
+impl Default for Thread {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Thread {
+    /// Create a new, empty instance of `Thread`.
+    pub const fn new() -> Self {
+        Self {
+            dataset: OctetsOwned { vec: Vec::new() },
+        }
+    }
+
+    /// Return an in-place initializer for an empty `Thread` insrtance.
+    pub fn init() -> impl Init<Self> {
+        init!(Self {
+            dataset <- OctetsOwned::init(),
+        })
+    }
+
+    /// Get the Extended PAN ID from the operational dataset
+    pub fn dataset_ext_pan_id(dataset_tlv: &[u8]) -> Result<&[u8], Error> {
+        ThreadTLV::new(dataset_tlv).ext_pan_id()
+    }
+
+    /// Get the Extended PAN ID from the operational dataset
+    pub fn ext_pan_id(&self) -> &[u8] {
+        unwrap!(Self::dataset_ext_pan_id(&self.dataset.vec))
```

Comment:
The comment could state that this convenience version of the function might panic

Reply:
Hm, **actually no**.
If the peer-provided thread dataset TLV did not contain an extended PAN ID in the first place, it would've NOT be persisted as a `WirelessNetwork` in the first place, so having an infallible API here is OK.

See `WirelessCreds::id`.

So I would rather document (with internal doc), that the `unwrap!` should never fail. I.e. it is `unreachable!` in fact.

Reply:
If it is not clear, we are persisting the thread dataset directly encoded in TLV (in the thread version of the TLV format, that is) because this way it occupies less space. Expanding it to its constituents would mean it would be much larger than 256 bytes, as then we need to use a bunch of heapless Vecs and Strings, which would be half-empty most of the time.

We do similar trick with the Matter certs, which are kept directly in their matter-TLV-encoded format, which consumes quite a bit less space.